clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java

```diff
@@ -102,53 +106,97 @@
         new Field(PARTITIONS_KEY_NAME, new ArrayOf(Type.INT32)));
 
     public static final Schema SUBSCRIPTION_V0 = new Schema(
-            new Field(TOPICS_KEY_NAME, new ArrayOf(Type.STRING)),
-            new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
+        new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES),
+        new Field(TOPICS_KEY_NAME, new ArrayOf(Type.STRING)));
 
     public static final Schema SUBSCRIPTION_V1 = new Schema(
-        new Field(TOPICS_KEY_NAME, new ArrayOf(Type.STRING)),
         new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES),
+        new Field(TOPICS_KEY_NAME, new ArrayOf(Type.STRING)),
         new Field(OWNED_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)));
 
     public static final Schema ASSIGNMENT_V0 = new Schema(
-            new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
-            new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
+        new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES),
+        new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)));
 
     public static final Schema ASSIGNMENT_V1 = new Schema(
-        new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
         new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES),
-        ERROR_CODE);
+        new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)));
 
-    public enum AssignmentError {
-        NONE(0),
-        NEED_REJOIN(1);
 
-        private final short code;
+    public static ByteBuffer serializeSubscription(Subscription subscription, Short version) {
+        switch (version) {
+            case CONSUMER_PROTOCOL_V0:
+                return serializeSubscriptionV0(subscription);
+
+            case CONSUMER_PROTOCOL_V1:
+                return serializeSubscriptionV1(subscription);
 
-        AssignmentError(final int code) {
-            this.code = (short) code;
+            default:
+                // for any versions higher than known, try to serialize it as V1
+                return serializeSubscriptionV1(subscription);
         }
+    }
 
-        public short code() {
-            return code;
+    public static Subscription deserializeSubscription(ByteBuffer buffer) {
```

Comment:
I removed the version number entirely from any class since it seems like the assignor (or consumer) actually shouldn't care -- if on the old version, ownedPartitions will just be initialized to empty list which is what it would be anyways. 
Also, the Assignment doesn't even need a version bump if we remove the error code, but I thought maybe we should bump it anyways just to keep it aligned with the Subscription version? Or is that unnecessary

Reply:
> Also, the Assignment doesn't even need a version bump if we remove the error code

The version of subscription / assignment should be aligned as a way to communicate the version of the client as well. So even without new fields in Assignment we should still bump up the version.